tasks/update_branches.py

```diff
@@ -10,103 +10,77 @@
     async def run_async(
-        self, db_session, *args, branch_name=None, ownerid=None, dry_run=True, **kwargs
+        self,
+        db_session,
+        *args,
+        branch_name=None,
+        incorrect_commitid=None,
+        dry_run=True,
+        **kwargs
     ):
         if branch_name is None:
             log.warning("No branch name specified, not updating any branches")
             return {"attempted": False}
 
         log.info(
             "Doing update branches for branch",
-            extra=dict(branch_name=branch_name, ownerid=ownerid),
+            extra=dict(branch_name=branch_name),
         )
-        if ownerid is not None:
-            log.info(
-                "Owner id was specified, only updating branches in the repo of that owner",
-                extra=dict(branch_name=branch_name, ownerid=ownerid),
-            )
-            repoids = (
-                db_session.query(Repository.repoid)
-                .filter(Repository.ownerid == ownerid)
-                .all()
-            )
-            log.info(
-                "repo ids we're taking a look at",
-                extra=dict(repoids=repoids, branch_name=branch_name, ownerid=ownerid),
-            )
-            query = (
-                db_session.query(Branch)
-                .filter(Branch.branch == branch_name, Branch.repoid.in_(repoids))
-                .yield_per(10)
-            )
-        else:
-            log.info(
-                "No owner id specified updating for branches in all orgs' repos",
-                extra=dict(branch_name=branch_name, ownerid=ownerid),
+
+        branches_to_update = (
+            db_session.query(Branch)
+            .filter(
+                Branch.branch == branch_name,
+                Branch.head == incorrect_commitid,
             )
-            query = (
-                db_session.query(Branch)
-                .filter(
-                    Branch.branch == branch_name,
+            .all()
+        )
+
+        chunk_size = 1000
+        chunks = [
+            branches_to_update[i : i + chunk_size]
+            for i in range(len(branches_to_update), chunk_size)
+        ]
+
+        for chunk in chunks:
+            for branch in chunk:
+                log.info(
+                    "Updating branch on repo",
+                    extra=dict(branch_name=branch_name, repoid=branch.repoid),
                 )
-                .yield_per(10)
-            )
 
-        for branch in query:
-            log.info(
-                "Updating branch on repo",
-                extra=dict(branch_name=branch_name, repoid=branch.repoid),
-            )
-            existing_commit = (
-                db_session.query(Commit)
-                .filter(Commit.repoid == branch.repoid, Commit.commitid == branch.head)
-                .first()
-            )
-            if existing_commit is not None:
+                latest_commit_on_branch = (
+                    db_session.query(Commit)
```

Comment:
This query will run _many_ times.

If we want to optimize to run the least number of queries in the DB as possible (as to now disturb normal operations) I think this should be refactored.
The trade-off as you'd expect is use more memory in the task to pre-compute certain values. 

My recommendation, if possible (haven't looked into it but I think it should be) is to get the latest commit for _all_ the `repoid` in the chunk once before the start of the loop and use that info to make the updates.

something like 
```
SELECT Commit WHERE branch == branch_name, repoid IN (repoids) GROUP BY repoid -picking last row by updatetimestamp- 
```

The part I don't know exactly how to do is the `-picking last row by updatetimestamp- ` but I think it should be possible to do that...

What do you think about that? Assuming to make the SELECT you need to possibly wait for locks maybe it's faster in the aggregate to do this.